ACPs/99-validatorsetmanager-contract/README.md

```diff
@@ -0,0 +1,370 @@
+
+Given these assumptions and the fact that most of the active blockchains on Avalanche mainnet are EVM-based, we propose defining a reference implementation for a Solidity smart contract that can:
+
+1. Hold relevant information about the current Subnet validator set, as well as historical information
+2. Send validator set updates to the P-Chain by generating `AdressedCall`s defined in ACP-77
+3. Correctly update the validator set by interpreting notification messages received from the P-Chain
+4. Be easily extended with custom “security modules” to implement any security model (e.g. PoS). Those modules have to implement the `IACP99SecurityModule` interface and will be called by the `ACP99Manager` contract upon validator set updates.
+
+Having an audited and open-source reference implementation freely available will contribute to lowering the cost of launching Subnets on Avalanche.
+
+Once deployed, the `ACP99Manager` contract will be used as the `Address` in the [`ConvertSubnetTx`](https://github.com/avalanche-foundation/ACPs/tree/main/ACPs/77-reinventing-subnets#convertsubnettx).
+
+## Specification
+
+**Note:** The naming convention followed for the interfaces and contracts are inspired from the way [OpenZeppelin Contracts](https://docs.openzeppelin.com/contracts/5.x/) are named after ERC standards, using `ACP` instead of `ERC`.
+
+### IACP99Manager
+
+Here is the proposed interface for the `IACP99Manager` contract:
+
+```solidity
+interface IACP99Manager {
+    /// @notice Subnet validation status
+    enum ValidationStatus {
+        Registering,
+        Active,
+        Updating,
+        Removing,
+        Completed,
+        Expired
+    }
+
+    /**
+     * @notice Subnet validation
+     * @param status The validation status
+     * @param nodeID The NodeID of the validator
+     * @param startTime The start time of the validation
+     * @param endTime The end time of the validation
+     * @param periods The list of validation periods.
+     * The index is the nonce associated with the weight update.
+     * @param activeSeconds The time during which the validator was active during this validation
+     * @param uptimeSeconds The uptime of the validator for this validation
+     */
+    struct Validation {
+        ValidationStatus status;
+        bytes32 nodeID;
+        uint64 startTime;
+        uint64 endTime;
+        ValidationPeriod[] periods;
+        uint64 activeSeconds;
+        uint64 uptimeSeconds;
+    }
+
+    /**
+     * @notice Subnet validation period
+     * @param weight The weight of the validator during the period
+     * @param startTime The start time of the validation period
+     * @param endTime The end time of the validation period (only ≠ 0 when the period is over)
+     */
+    struct ValidationPeriod {
+        uint64 weight;
+        uint64 startTime;
+        uint64 endTime;
+    }
```

Comment:
- Yes, it is possible to use `validation.periods[0].startTime` instead of `validation.startTime` to avoid duplicating data. For `endTime` it would be a little bit trickier: `validation.periods[validation.periods.length - 1].endTime` (but irrelevant to store `endTime` anyway if we choose to remove history
- I thought it was enough to use the array index as the place to store the nonce since it is incremented by 1 at each weight change
- I chose to track `uptimeSeconds` at the `Validation` level because I don't think it is useful to track it at the `ValidationPeriod` level and thought it might be difficult to differentiate the uptime from a specific period than from the whole validation. The latter depends on how the uptime is tracked at the VM level. Do we have more information about how this will be handled btw?

About the validation historic tracking, I agree that it is most likely not needed. Thus my open question at the end.